src/track/beats.cpp

```diff
@@ -1,29 +1,158 @@
-
 #include "track/beats.h"
 
+#include "track/beatutils.h"
+#include "track/track.h"
+
 namespace mixxx {
 
-int Beats::numBeatsInRange(double dStartSample, double dEndSample) {
-    double dLastCountedBeat = 0.0;
+const QString Beats::BEAT_MAP_VERSION = "BeatMap-1.0";
+const QString Beats::BEAT_GRID_1_VERSION = "BeatGrid-1.0";
+const QString Beats::BEAT_GRID_2_VERSION = "BeatGrid-2.0";
+
+namespace {
+inline bool BeatLessThan(const track::io::Beat& beat1, const track::io::Beat& beat2) {
+    return beat1.frame_position() < beat2.frame_position();
+}
+
+inline FrameDiff_t getBeatLengthFrames(Bpm bpm, double sampleRate) {
+    return 60.0 * sampleRate / bpm.getValue();
+}
+} // namespace
+
+Beats::Beats(const Track* track, const QVector<FramePos>& beats)
+        : Beats(track) {
+    if (beats.size() > 0) {
+        FramePos previous_beatpos = FramePos(-1);
+        track::io::Beat protoBeat;
+
+        for (auto beat : beats) {
+            beat.setValue(beat.getValue());
+            if (beat <= previous_beatpos || beat < FramePos(0)) {
```

Comment:
What is the solution if the first beat is cut for some reasons and is unbearable at the pre-roll?